### PR TITLE
Fix mess if a class declared @property but did not declare @cclass

### PR DESCRIPTION
**Pull request by shrinktofit**
### Changelog

* Fix mess if a class declared `@property` but did not declare `@cclass` . This closes https://github.com/cocos/3d-tasks/issues/8697 . See linked issue for detail.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: tests/core/decorator.test.ts [hunk 1]
```diff
@@ -381,4 +381,65 @@ describe('Decorated property test', () => {
...
+    @ccclass
+    class C extends A {
+        @property c = 0;
+    }
+
+    expect(CCClass.isCCClassOrFastDefined(A)).toBe(false);
+    expect(getProps(A)).toBeUndefined();
+
+    expect(CCClass.isCCClassOrFastDefined(B)).toBe(true);
+    expect(getProps(B)).toStrictEqual(['b']);
```
Comment (shrinktofit) [inline tests/core/decorator.test.ts:433]: Before: `['a', 'b']`

File: tests/core/decorator.test.ts [hunk 2]
```diff
@@ -381,4 +381,65 @@ describe('Decorated property test', () => {
...
+
+    expect(CCClass.isCCClassOrFastDefined(C)).toBe(true);
+    expect(getProps(C)).toStrictEqual(['c']);
```
Comment (shrinktofit) [inline tests/core/decorator.test.ts:436]: Before: `['a', 'c', 'b']`

File: cocos/core/data/decorators/utils.ts [hunk 3]
```diff
@@ -144,16 +145,28 @@ export function makeSmartEditorClassDecorator<TValue> (propertyName: string, def
     });
 }
 
-// caches for class construction
-export const CACHE_KEY = '__ccclassCache__';
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (shrinktofit) [inline cocos/core/data/decorators/utils.ts]:
```suggestion
```

File: cocos/core/data/decorators/utils.ts [hunk 4]
```diff
@@ -26,6 +26,7 @@ import { DEV } from 'internal:constants';
 import { CCClass } from '../class';
 import { error } from '../../platform/debug';
 import { getClassName } from '../../utils/js-typed';
+import { ClassStash } from '../class-stash';
```
Comment (jareguo) [inline cocos/core/data/decorators/utils.ts]: import type?

**Review by jareguo — APPROVED** [review]

File: cocos/core/data/decorators/utils.ts (hunk 4 above)
Comment (shrinktofit) [reply cocos/core/data/decorators/utils.ts]: OK: https://github.com/cocos/cocos-engine/pull/15652/commits/764ff437ff76a7ffa87254eae571edde24a113aa

**Review by minggo — APPROVED** [review]

**Pull request merged**